### ISSUE TITLE
API feedback - global state

### ISSUE THREAD

**Issue by rbtcollins** [PagerDuty/pagerduty-api-python-client#13]
Hi, I just wanted to provide a couple of bits of feedback having given this a bit of exercise this week.

Firstly, global state is ugh: pypd.api_key = <thing>, followed by pypd.User.find() or whatever: this makes it very hard and fragile to write code that talks to more than one pagerduty account within the same process. Thats avoidable with one of the other idioms you can see around the Python ecosystem:
```
api_client = pypd.APIClient(key=...)
api_client.User.find()
```

or
```
api_client = pypd.APIClient(key=...)
pypd.User(api_client).find()
```

or 
````
api_client = pypd.APIClient(key=...)
pypd.User().find(api_client)
```

Basically in the first case, pypd would track any state itself, in the latter two the user supplies the api client in a late bound fashion - which is equally convenient, because its still just one thing that needs to be passed around within a client program. I don't really have an opinion about whats best, just please - not global state.

**Comment by jdcumpson** [comment]
Yeah this is something I hate too - I did it because traditionally it makes the 5-second how-do adoption rate "way easier". I much prefer clients to implicit variables. However, that being said I still wanted to make it feel very imperative and simple. I'll think about this feedback if I do another API client from scratch. It wasn't the base use-case so I ignored it as long as no one complained. Thanks for complaining 😆 

I think in order to not break backwards compat and be nice like we'd both like the client-style solution would be best.. I used keyspace because I think clients ought to be handling connections and this name made more sense right now.
```
foo_keyspace = pypd.Keyspace(key='foo`)
bar_keyspace = pypd.Keyspace(key='bar')
pypd.api_key = 'other_secret'

# n-client namespaces can be generated
foo_user = foo_keyspace.User.find_one(...)
assert foo_user.api_key == 'foo'

bar_user = bar_keyspace.User.find_one(...)
assert bar_user.api_key == 'bar'

# don't break backwards compat
pypd_user = pypd.User.find_one(...)
assert pypd_user.api_key == 'secret'
```

I think it would be a pretty trivial class 
```
class Keyspace(object):
    api_key = None
    
    def __init__(self, api_key=None):
        # some meta programming magic
        self.api_key = api_key if api_key else pypd.api_key
        for model_class_name in [...]:
            klass = getattr(__import__('pypd.{}'.format(model_class_name), model_class_name)
            keyspaced_class = type(model_class_name, (klass, ), {'api_key': api_key})
            setattr(model_class_name, keyspaced_class)
```

**Comment by jdcumpson** [comment]
@rbtcollins I just realized something that I totally missed that day, if you look at the test cases - `Entity.find(...)` and `Entity.find_one(...)` both accept the keyword argument `api_key` where you can do something like `Entity.find(api_key='MYKEY')`. While not great, I remember adding this backdoor for just this reason. Can you let me know if this is sufficient for now?

**Comment by rbtcollins** [comment]
I think we if change the docs to show the safe way of doing it (whether in addition to or instead of the global state is moot), then it would be much more discoverable.

**Comment by jdcumpson** [comment]
Cool, then the easy task to close out this issue is making some more inclusive docs. Let me know if that's what you've meant.

**Comment by bzmw** [comment]
Hey there, 
Thanks for your contribution but we decided to deprecate this project. Please use [PDPyras](https://github.com/PagerDuty/pdpyras) instead.
